include/device_uid.h

```diff
@@ -0,0 +1,87 @@
+/**
+ * @file
+ *
+ * @brief Public APIs to get device Unique ID.
+ */
+
+/*
+ * Copyright (c) 2018 Alexander Wachter
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DEVICE_UID_H_
+#define ZEPHYR_INCLUDE_DEVICE_UID_H_
+
+/**
+ * @brief Device Unique ID
+ * @{
+ */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <zephyr/types.h>
+#include <device.h>
+
+/**
+ * @brief Copy the device unique id to a buffer
+ *
+ * This routine copies "length" number of bytes of the device UID to the buffer.
+ * If the device UID is smaller then length, the rest is left unchanged.
+ *
+ * @param dev     Pointer to the device structure for the driver instance.
+ * @param buffer  Buffer to write the ID to.
+ * @param length  Max length of the buffer.
+ *
+ * @retval size of the device UID copied.
+ */
+typedef size_t (*device_uid_get_t)(struct device *dev,
+				   u8_t *buffer, size_t length);
+
+/**
+ * @brief Get the length of the deice UID
+ *
+ * Get the length of the deice UID in byte
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ *
+ * @retval length of the device UID
+ */
+typedef size_t (*device_uid_get_length_t)(struct device *dev);
+
+struct device_uid_api {
+	device_uid_get_t get_uid;
+	device_uid_get_length_t get_length;
+};
+
+
+__syscall size_t device_uid_get(struct device *dev, u8_t *buffer, size_t length);
+
+static inline size_t _impl_device_uid_get(struct device *dev, u8_t *buffer,
```

Comment:
What about returning the length when `buffer` is `NULL`? We'd save one API call that way:

```
int len = device_uid_get(dev, NULL, 0);
ret = device_uid_get(dev, &buf, len);
```

Reply:
Then using this API as additional entropy source, this could be dangerous.

Reply:
Why? Not sure I understand

Reply:
Because its confusing that it returns a length even the length passed was zero. For example the rest of a pre filled array should be filled with the device id but the rest is zero. In my opinion this is more confusing than have two API calls.

Reply:
Then make it work the usual POSIX way: let user pass buffer of arbitrary length, and return number of bytes actually written there.

Reply:
That's how it actually works.

Reply:
Done

Reply:
Sorry, not sure about what "Done" means here. It's still 2 system calls right?

Reply:
Nope. static inline size_t device_uid_get_length(struct device *dev) is not a syscall

Reply:
if device id is fixed length for given device we might want to have define (kconfig or dts). Something like:
```
u8_t dev_uid[CONFIG_DEVICE_UID_LENGTH];
device_uid_get(dev, dev_uid, CONFIG_DEVICE_UID_LENGTH);
```